rasa_core/policies/embedding_policy.py

```diff
@@ -0,0 +1,848 @@
+                    feed_dict={self.a_in: batch_a,
+                               self.b_in: batch_b,
+                               self.c_in: batch_c,
+                               self.is_training: True}
+                )
+                ep_loss += sess_out.get('loss') / batches_per_epoch
+
+            pbar.set_postfix({"loss": ep_loss})
+
+            if logger.isEnabledFor(logging.INFO) and (
+                    (ep + 1) % 50 == 0 or (ep + 1) == self.epochs):
+                self._output_training_stat(X, extras,
+                                           actions_for_X, all_Y_d,
+                                           mask, ep, ep_loss)
+
+    def _output_training_stat(self,
+                              X, extras, actions_for_X, all_Y_d,
+                              mask, ep, ep_loss):
+        """Output training statistics"""
+        n = 100  # choose n examples to calculate train accuracy
+        ids = np.random.permutation(len(X))[:n]
+        all_Y_d_x = np.stack([all_Y_d for _ in range(X[ids].shape[0])])
+
+        _sim, _mask = self.session.run(
+            [self.sim_op, mask],
+            feed_dict={self.a_in: X[ids],
+                       self.b_in: all_Y_d_x,
+                       self.c_in: extras[ids],
+                       self.is_training: False}
+        )
+
+        train_acc = np.sum((np.argmax(_sim, -1) ==
+                            actions_for_X[ids]) * _mask)
+        train_acc /= np.sum(_mask)
+        logger.info("epoch {} / {}: loss={:.3f}, accuracy={:.3f}"
+                    "".format((ep + 1), self.epochs,
+                              ep_loss, train_acc))
+
+    def train(self,
+              training_trackers,  # type: List[DialogueStateTracker]
+              domain,  # type: Domain
+              **kwargs  # type: **Any
+              ):
+        # type: (...) -> None
+        """Trains the policy on given training trackers."""
+
+        # dealing with training data
+        logger.debug('Started to train embedding policy.')
+
+        training_data = self.featurize_for_training(training_trackers,
+                                                    domain,
+                                                    **kwargs)
+        if kwargs:
+            self.config.update(kwargs)
+            logger.debug("Config is updated with {}".format(kwargs))
+            self._load_params()
+
+        self.mean_time = np.mean(training_data.true_length)
+
+        self.encoded_all_actions = self._create_encoded_actions(domain)
+
+        # check if number of negatives is less than number of actions
+        logger.debug("Check if num_neg {} is smaller "
+                     "than number of actions {}, "
+                     "else set num_neg to the number of actions - 1"
+                     "".format(self.num_neg, domain.num_actions))
+        self.num_neg = min(self.num_neg, domain.num_actions - 1)
+
+        X, Y, extras, actions_for_X = self._prepare_data_for_training(
```

Comment:
there is a lot of data munging going on - are there any tests to make sure the representation during training is the same as during prediction?

Reply:
I call the same function to create X during training and prediction

Reply:
but yeah, I guess it is still good, to create a test for it